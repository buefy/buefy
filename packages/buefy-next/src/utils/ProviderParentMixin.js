import { hasFlag } from './helpers'

const items = 1
const sorted = 3

export const Items = items
export const Sorted = sorted

export default (itemName, flags = 0) => {
    const mixin = {
        provide() {
            return {
                ['b' + itemName]: this
            }
        }
    }

    if (hasFlag(flags, items)) {
        mixin.data = function () {
            return {
                childItems: [],
                ...(hasFlag(flags, sorted) ? { nextIndex: 0 } : {})
            }
        }
        mixin.methods = {
            _registerItem(item) {
                if (hasFlag(flags, sorted)) {
                    // assigns a dynamic index.
                    // dynamic indices will be messed up if any child is
                    // unmounted.
                    // use the new `order` prop to maintain the ordering.
                    item.dynamicIndex = this.nextIndex
                    ++this.nextIndex
                }
                this.childItems.push(item)
            },
            _unregisterItem(item) {
                this.childItems = this.childItems.filter((i) => i.value !== item.value)
            }
        }

        if (hasFlag(flags, sorted)) {
            mixin.computed = {
                /**
                 * When items are added/removed sort them according to their position
                 */
                sortedItems() {
                    return this.childItems.slice().sort((i1, i2) => {
                        return i1.index - i2.index
                    })
                }
            }
        }
    }
    return mixin
}
