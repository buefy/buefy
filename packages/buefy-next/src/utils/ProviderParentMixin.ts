import type { DefineComponent } from 'vue'

import { hasFlag } from './helpers'
import type { InjectedChild } from './ProvideInjectTypes'

const items = 1
const sorted = 3

export const Items = items
export const Sorted = sorted

export type FeatureFlags = 0 | typeof Items | typeof Sorted

/* eslint-disable @typescript-eslint/ban-types */
// conditional data fields of ProviderParentMixin
export type ProviderParentMixinData<
    Flags extends FeatureFlags,
    Child extends InjectedChild,
    HasItems extends boolean = Flags extends typeof Items | typeof Sorted ? true : false,
    IsSorted extends boolean = Flags extends typeof Sorted ? true : false
> = (HasItems extends true ? { childItems: Child[] } : {})
    & (IsSorted extends true ? { nextIndex: number } : {})

// conditional computed fields of ProviderParentMixin
export type ProviderParentMixinComputed<
    Flags extends FeatureFlags,
    Child extends InjectedChild,
    IsSorted extends boolean = Flags extends typeof Sorted ? true : false
> = IsSorted extends true ? { sortedItems(): Child[] } : {}

// conditional methods of ProviderParentMixin
export type ProviderParentMixinMethods<
    Flags extends FeatureFlags,
    Child extends InjectedChild,
    HasItems extends boolean = Flags extends typeof Items | typeof Sorted ? true : false
> = HasItems extends true ? {
    _registerItem(item: Child): void
    _unregisterItem(item: Child): void
} : {}

export type ProviderParentMixin<
    Flags extends FeatureFlags,
    Child extends InjectedChild
> = DefineComponent<
    {}, // PropsOrPropOptions
    {}, // RawBindings
    ProviderParentMixinData<Flags, Child>, // D(ata)
    ProviderParentMixinComputed<Flags, Child>, // C(computed)
    ProviderParentMixinMethods<Flags, Child> // M(ethods)
>
/* eslint-enable @typescript-eslint/ban-types */

export default <
    Flags extends FeatureFlags = FeatureFlags,
    Child extends InjectedChild = InjectedChild
>(itemName: string, flags?: Flags): ProviderParentMixin<Flags, Child> => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const mixin: any = {
        provide() {
            return {
                ['b' + itemName]: this
            }
        }
    }

    if (flags !== undefined && hasFlag(flags, items)) {
        mixin.data = function () {
            return {
                childItems: [] as Child[],
                ...(hasFlag(flags, sorted) ? { nextIndex: 0 } : {})
            }
        }
        mixin.methods = {
            _registerItem(item: Child) {
                if (hasFlag(flags, sorted)) {
                    // assigns a dynamic index.
                    // dynamic indices will be messed up if any child is
                    // unmounted.
                    // use the new `order` prop to maintain the ordering.
                    item.dynamicIndex = this.nextIndex
                    ++this.nextIndex
                }
                this.childItems.push(item)
            },
            _unregisterItem(item: Child) {
                this.childItems = this.childItems
                    .filter((i: Child) => i.uniqueValue !== item.uniqueValue)
            }
        }

        if (hasFlag(flags, sorted)) {
            mixin.computed = {
                /**
                 * When items are added/removed sort them according to their position
                 */
                sortedItems() {
                    return this.childItems.slice().sort((i1: Child, i2: Child) => {
                        return i1.index - i2.index
                    })
                }
            }
        }
    }
    return mixin
}
