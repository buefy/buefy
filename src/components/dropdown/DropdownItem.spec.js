import { shallowMount } from '@vue/test-utils'
import BDropdownItem from '@components/dropdown/DropdownItem'
import ProviderParentMixin from '../../utils/ProviderParentMixin'

let wrapper
const dropdownSelected = 'val'
const BDropdown = {
    mixins: [ProviderParentMixin('dropdown')],
    template: '<b-dropdown-stub></b-dropdown-stub>',
    data() {
        return {
            selected: dropdownSelected
        }
    }
}

describe('BDropdownItem', () => {
    beforeEach(() => {
        wrapper = shallowMount(BDropdownItem, {
            parentComponent: BDropdown
        })
    })

    it('is called', () => {
        expect(wrapper.name()).toBe('BDropdownItem')
        expect(wrapper.isVueInstance()).toBeTruthy()
    })

    it('render correctly', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('returns item classes accordingly', () => {
        const value = dropdownSelected
        const hasLink = true
        const disabled = false
        const paddingless = true
        const isActive = true
        wrapper.setProps({
            value,
            hasLink,
            disabled,
            paddingless
        })
        expect(wrapper.vm.itemClasses).toEqual({
            'dropdown-item': !hasLink,
            'is-disabled': disabled,
            'is-paddingless': paddingless,
            'is-active': isActive,
            'has-link': hasLink
        })
    })

    it('returns if item is clickable accordingly', () => {
        expect(wrapper.vm.isClickable).toBeTruthy()

        wrapper.setProps({ separator: true })
        expect(wrapper.vm.isClickable).toBeFalsy()
    })

    it('react accordingly when the item is selected', () => {
        wrapper.vm.$parent.selectItem = jest.fn()

        wrapper.setProps({ separator: true })
        wrapper.vm.selectItem()
        expect(wrapper.vm.$parent.selectItem).not.toHaveBeenCalled()

        wrapper.setProps({ separator: false })
        wrapper.vm.selectItem()
        expect(wrapper.vm.$parent.selectItem).toHaveBeenCalled()
        expect(wrapper.emitted()['click']).toBeTruthy()
    })
})
