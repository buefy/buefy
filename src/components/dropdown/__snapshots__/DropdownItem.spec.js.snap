// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BDropdownItem render correctly 1`] = `<a tabindex="0" class="dropdown-item"></a>`;
